### PR TITLE
JP-3638: Flag asymmetrical snowballs

### PR DESCRIPTION
**Pull request by mwregan2**
Resolves [JP-3638](https://jira.stsci.edu/browse/JP-3638)

Closes spacetelescope/jwst#8516

This PR addresses the problem of asymmetrical snowballs not getting removed from NIR exposures. These are a sub-population of snowballs where the incoming particle leads a trail of pixels that are flagged as jump and then trigger a circular snowball with a saturated core. Because the current code requires that the center of the ellipse fitted to the pixel flagged as jump have a saturated pixel, these snowballs do not get recognized as a snowball and do not get any increase in the jump ellipse. 

The change was to just remove the requirement that the central pixel be saturated and change the requirement to be that there needs to be saturation within a circle with a radius of the major axis of the ellipse centered on the center of the ellipse. 

**Checklist**

- [x] added entry in `CHANGES.rst` (either in `Bug Fixes` or `Changes to API`)
- [x] updated relevant tests
- [ ] updated relevant documentation
- [ ] updated relevant milestone(s)
- [x] added relevant label(s)

### CODE REVIEW

File: src/stcal/jump/jump.py [hunk 1]
```diff
@@ -465,6 +465,7 @@ def detect_jumps(
 
         #  This is the flag that controls the flagging of snowballs.
         if expand_large_events:
+#            fits.writeto("ingdq.fits", gdq)
```
Comment (kmacdonald-stsci) [inline src/stcal/jump/jump.py:468]: Should the pipeline be outputting files that aren't expected?

File: tests/test_jump.py [hunk 2]
```diff
@@ -545,10 +550,11 @@ def test_inside_ellipes5():
 @pytest.mark.skip(" used for local testing")
 def test_flag_persist_groups():
 #   gdq = fits.getdata("persistgdq.fits")
-    gdq = np.zeros(shape=(2,2,2,2))
-    print(gdq.shape[0])
-    gdq = gdq[:, 0:10, :, :]
-    total_snowballs = flag_large_events(
+    gdq = fits.getdata("../../../notebooks/ingdq.fits")
```
Comment (kmacdonald-stsci) [inline tests/test_jump.py:553]: This should not be here.  The 'ingdq.fits' file should not be outputted by the pipeline.  And for testing, all files should be written to the temporary directory.

File: tests/test_jump.py [hunk 3]
```diff
@@ -561,7 +567,7 @@ def test_flag_persist_groups():
         sat_expand=1.1,
         mask_persist_grps_next_int=True,
         persist_grps_flagged=0)
-
+    fits.writeto("../../../notebooks/gdqout.fits", gdqout, overwrite=True)
```
Comment (kmacdonald-stsci) [inline tests/test_jump.py:570]: Tests should only write pipeline files.  And all files should be written to the `pytest` directory.

File: src/stcal/ramp_fitting/ols_fit.py [hunk 4]
```diff
@@ -9,7 +9,7 @@
 
 import numpy as np
 
-from .slope_fitter import ols_slope_fitter  # c extension
+#from .slope_fitter import ols_slope_fitter  # c extension
```
Comment (kmacdonald-stsci) [inline src/stcal/ramp_fitting/ols_fit.py:12]: Why is this commented out?

**Review by kmacdonald-stsci — CHANGES_REQUESTED** [review]
The pipeline should only output expected output files.  All output files should be written to the `pytest` temporary folder.  The imports in the `ols_fit.py` file should not be commented out.

File: tests/test_jump.py (hunk 2 above)
Comment (mwregan2) [reply tests/test_jump.py:553]: The test is set to skip so I don't understand how it is a problem but I'll remove the test.

File: src/stcal/ramp_fitting/ols_fit.py (hunk 4 above)
Comment (mwregan2) [reply src/stcal/ramp_fitting/ols_fit.py:12]: This causes an unresolved reference for both .slope_fitts and ols_slope_fitter. I've removed the comment.

File: tests/test_jump.py (hunk 3 above)
Comment (mwregan2) [reply tests/test_jump.py]: I removed the write but it is in a test that is set to skip. Removing these tests means I have to rewrite the tests every time I want to test without having to run the full step.
If there's a better way to do this, I open to suggestions.

**Comment by mwregan2** [comment]
Ken,
  I’m getting a test failure on the test
test / test_downstream / py311-jwst-cov-xdist (ubuntu-latest) (pull_request)
  I see two “F” values in the results but I’ve never been able to figure out how to map the “F”values to the individual test. Is there something that would tell me what failed?
Thanks
Mike


From: Ken MacDonald ***@***.***>
Reply-To: spacetelescope/stcal ***@***.***>
Date: Monday, June 3, 2024 at 4:52 PM
To: spacetelescope/stcal ***@***.***>
Cc: Michael Regan ***@***.***>, Author ***@***.***>
Subject: Re: [spacetelescope/stcal] Flag asymmetrical snowballs (PR #261)


@kmacdonald-stsci requested changes on this pull request.

The pipeline should only output expected output files. All output files should be written to the pytest temporary folder. The imports in the ols_fit.py file should not be commented out.

________________________________

In src/stcal/jump/jump.py<https://urldefense.com/v3/__https:/github.com/spacetelescope/stcal/pull/261*discussion_r1625028768__;Iw!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!y32R9rWbrWl4BT7hAOdcC6SbsycTHQENRtm8mAqxT62_RBtAbFBP_Kt7CPnwPv0rAw8XLEAwee88Ey9n8gnJbsU$>:

> @@ -465,6 +465,7 @@ def detect_jumps(



         #  This is the flag that controls the flagging of snowballs.

         if expand_large_events:

+#            fits.writeto("ingdq.fits", gdq)

Should the pipeline be outputting files that aren't expected?

________________________________

In tests/test_jump.py<https://urldefense.com/v3/__https:/github.com/spacetelescope/stcal/pull/261*discussion_r1625030946__;Iw!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!y32R9rWbrWl4BT7hAOdcC6SbsycTHQENRtm8mAqxT62_RBtAbFBP_Kt7CPnwPv0rAw8XLEAwee88Ey9nF1McL1k$>:

> @@ -545,10 +550,11 @@ def test_inside_ellipes5():

 @pytest.mark.skip(" used for local testing")

 def test_flag_persist_groups():

 #   gdq = fits.getdata("persistgdq.fits")

-    gdq = np.zeros(shape=(2,2,2,2))

-    print(gdq.shape[0])

-    gdq = gdq[:, 0:10, :, :]

-    total_snowballs = flag_large_events(

+    gdq = fits.getdata("../../../notebooks/ingdq.fits")

This should not be here. The 'ingdq.fits' file should not be outputted by the pipeline. And for testing, all files should be written to the temporary directory.

________________________________

In tests/test_jump.py<https://urldefense.com/v3/__https:/github.com/spacetelescope/stcal/pull/261*discussion_r1625032060__;Iw!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!y32R9rWbrWl4BT7hAOdcC6SbsycTHQENRtm8mAqxT62_RBtAbFBP_Kt7CPnwPv0rAw8XLEAwee88Ey9nizLD5pU$>:

> @@ -561,7 +567,7 @@ def test_flag_persist_groups():

         sat_expand=1.1,

         mask_persist_grps_next_int=True,

         persist_grps_flagged=0)

-

+    fits.writeto("../../../notebooks/gdqout.fits", gdqout, overwrite=True)

Tests should only write pipeline files. And all files should be written to the pytest directory.

________________________________

In src/stcal/ramp_fitting/ols_fit.py<https://urldefense.com/v3/__https:/github.com/spacetelescope/stcal/pull/261*discussion_r1625034818__;Iw!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!y32R9rWbrWl4BT7hAOdcC6SbsycTHQENRtm8mAqxT62_RBtAbFBP_Kt7CPnwPv0rAw8XLEAwee88Ey9nKMOsKFU$>:

> @@ -9,7 +9,7 @@



 import numpy as np



-from .slope_fitter import ols_slope_fitter  # c extension

+#from .slope_fitter import ols_slope_fitter  # c extension

Why is this commented out?

File: CHANGES.rst [hunk 5]
```diff
@@ -13,7 +13,11 @@ Changes to API
 
 Bug Fixes
 ---------
-
+jump
+~~~~
+- Flag asymmetrical snowballs that are missed by the current code (JP-3638). The was changed to
```
Comment (hbushouse) [inline CHANGES.rst:18]: `The was changed ...` needs fixing

**Review by hbushouse — COMMENTED** [review]
Needs a regression test run. I'll start one and leave a link here.

**Comment by hbushouse** [comment]
Regtest started at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1496

**Comment by hbushouse** [comment]
@nden Who's on the hook from Roman these days to run romancal tests for this PR?

File: CHANGES.rst (hunk 5 above)
Comment (mwregan2) [reply CHANGES.rst]: I fixed the wording

**Comment by hbushouse** [comment]
@mwregan2 The regression test run shows differences in the jump step results for a NIRISS and NIRSpec exposure. For the NIRISS data, the number of extended CRs flagged, as recorded in the EXTNCRS keyword, went up by a factor of 10 compared to before. In the NIRSpec exposure it went up by a factor of 2. Are these large increases expected due to the changes here and are they reasonable?

**Comment by mwregan2** [comment]
I can see getting a factor of 2 as a minimum. I’ve look at a sample of long exposures and see what count rate I get.

From: Howard Bushouse ***@***.***>
Reply-To: spacetelescope/stcal ***@***.***>
Date: Tuesday, June 4, 2024 at 10:33 PM
To: spacetelescope/stcal ***@***.***>
Cc: Michael Regan ***@***.***>, Mention ***@***.***>
Subject: Re: [spacetelescope/stcal] Flag asymmetrical snowballs (PR #261)


@mwregan2<https://urldefense.com/v3/__https:/github.com/mwregan2__;!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!y8SY-uTNr5mv4-PuZkaRLinYtLEQzvkk8N-9vG_WKJU_FLHYScofnYikorerBxjLp2gMF5J1NjmVuTc5gWmu09U$> The regression test run shows differences in the jump step results for a NIRISS and NIRSpec exposure. For the NIRISS data, the number of extended CRs flagged, as recorded in the EXTNCRS keyword, went up by a factor of 10 compared to before. In the NIRSpec exposure it went up by a factor of 2. Are these large increases expected due to the changes here and are they reasonable?

**Comment by mwregan2** [comment]
I looked at four 30 min NIRSpec observations. The snowball count with up by 50% for all four of them. That will affect anything that’s looking at the variance of a background or some other region.


From: Michael Regan ***@***.***>
Date: Wednesday, June 5, 2024 at 9:38 AM
To: spacetelescope/stcal ***@***.***>, spacetelescope/stcal ***@***.***>
Cc: Mention ***@***.***>
Subject: Re: [spacetelescope/stcal] Flag asymmetrical snowballs (PR #261)

I can see getting a factor of 2 as a minimum. I’ve look at a sample of long exposures and see what count rate I get.

From: Howard Bushouse ***@***.***>
Reply-To: spacetelescope/stcal ***@***.***>
Date: Tuesday, June 4, 2024 at 10:33 PM
To: spacetelescope/stcal ***@***.***>
Cc: Michael Regan ***@***.***>, Mention ***@***.***>
Subject: Re: [spacetelescope/stcal] Flag asymmetrical snowballs (PR #261)


@mwregan2<https://urldefense.com/v3/__https:/github.com/mwregan2__;!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!y8SY-uTNr5mv4-PuZkaRLinYtLEQzvkk8N-9vG_WKJU_FLHYScofnYikorerBxjLp2gMF5J1NjmVuTc5gWmu09U$> The regression test run shows differences in the jump step results for a NIRISS and NIRSpec exposure. For the NIRISS data, the number of extended CRs flagged, as recorded in the EXTNCRS keyword, went up by a factor of 10 compared to before. In the NIRSpec exposure it went up by a factor of 2. Are these large increases expected due to the changes here and are they reasonable?

**Comment by mwregan2** [comment]
I’ve figured out the cause of the large increase in the NIRISS Snowballs with this PR. The current NIRISS parameters are very conservative and only detect the medium to large snowballs. This causes the step to miss most of the small snowballs which is what most of them are.
I talked with David and he recommended I run a subset of the new regression tests for NIRISS exposures and compare with and without the PR.
I’m working on that now.


From: Howard Bushouse ***@***.***>
Reply-To: spacetelescope/stcal ***@***.***>
Date: Tuesday, June 4, 2024 at 10:33 PM
To: spacetelescope/stcal ***@***.***>
Cc: Michael Regan ***@***.***>, Mention ***@***.***>
Subject: Re: [spacetelescope/stcal] Flag asymmetrical snowballs (PR #261)


@mwregan2<https://urldefense.com/v3/__https:/github.com/mwregan2__;!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!y8SY-uTNr5mv4-PuZkaRLinYtLEQzvkk8N-9vG_WKJU_FLHYScofnYikorerBxjLp2gMF5J1NjmVuTc5gWmu09U$> The regression test run shows differences in the jump step results for a NIRISS and NIRSpec exposure. For the NIRISS data, the number of extended CRs flagged, as recorded in the EXTNCRS keyword, went up by a factor of 10 compared to before. In the NIRSpec exposure it went up by a factor of 2. Are these large increases expected due to the changes here and are they reasonable?

**Review by hbushouse — COMMENTED** [review]
Based on subsequent discussions regarding the large increase in snowball flagging for NIRISS images, it's been concluded that those are OK and actually desirable.

So my final question before approving is to ask whether this change should result in any updates in the docs anywhere?

**Comment by mwregan2** [comment]
I should update the report to mention this tweak to the algorithm. It really isn’t a fundamental change. It would just be nice to show an example of the offset saturation.


From: Howard Bushouse ***@***.***>
Reply-To: spacetelescope/stcal ***@***.***>
Date: Thursday, June 6, 2024 at 11:22 AM
To: spacetelescope/stcal ***@***.***>
Cc: Michael Regan ***@***.***>, Mention ***@***.***>
Subject: Re: [spacetelescope/stcal] JP-3638: Flag asymmetrical snowballs (PR #261)


@hbushouse commented on this pull request.

Based on subsequent discussions regarding the large increase in snowball flagging for NIRISS images, it's been concluded that those are OK and actually desirable.

So my final question before approving is to ask whether this change should result in any updates in the docs anywhere?

**Review by kmacdonald-stsci — APPROVED** [review]

**Review by hbushouse — APPROVED** [review]

**Pull request merged**